share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorOffsetsManager.java

```diff
@@ -0,0 +1,143 @@
+    // the values would automatically revert in accordance with the last committed offset.
+    private final TimelineLong minOffset;
+    // Initially true because we don't want to send the offset
+    // until findRedundantOffset determines a redundant offset is
+    // indeed present.
+    private AtomicBoolean offsetExposed = new AtomicBoolean(true);
+
+    public ShareCoordinatorOffsetsManager(SnapshotRegistry snapshotRegistry) {
+        Objects.requireNonNull(snapshotRegistry);
+        offsets = new TimelineHashMap<>(snapshotRegistry, 0);
+        minOffset = new TimelineLong(snapshotRegistry);
+        minOffset.set(Long.MAX_VALUE);  // For easy min update.
+    }
+
+    /**
+     * Method updates internal state with the supplied offset for the provided
+     * share partition key. It then calculates the minimum offset, if possible,
+     * below which all offsets are redundant.
+     *
+     * @param key    - represents {@link SharePartitionKey} whose offset needs updating
+     * @param offset - represents the latest partition offset for provided key
+     */
+    public void updateState(SharePartitionKey key, long offset) {
+        minOffset.set(Math.min(minOffset.get(), offset));
+        offsets.put(key, offset);
+
+        Optional<Long> deleteTillOffset = findRedundantOffset();
+        deleteTillOffset.ifPresent(off -> {
+            minOffset.set(off);
+            offsetExposed.set(false);
+        });
+    }
+
+    private Optional<Long> findRedundantOffset() {
+        if (offsets.isEmpty()) {
+            return Optional.empty();
+        }
+
+        long soFar = Long.MAX_VALUE;
+
+        for (long offset : offsets.values()) {
+            // Get min offset among latest offsets
+            // for all share keys in the internal partition.
+            soFar = Math.min(soFar, offset);
+
+            // minOffset represents the smallest necessary offset
+            // and if soFar equals it, we cannot proceed. This can happen
+            // if a share partition key hasn't had records written for a while
+            // For example,
+            // <p>
+            // key1:1
+            // key2:2 4 6
+            // key3:3 5 7
+            // <p>
+            // We can see in above that offsets 2, 4, 3, 5 are redundant,
+            // but we do not have a contiguous prefix starting at minOffset
+            // and we cannot proceed.
+            if (soFar == minOffset.get()) {
+                return Optional.empty();
+            }
+        }
+
+        return Optional.of(soFar);
+    }
+
+    /**
+     * Most recent last redundant offset. This method is to be used
+     * when the caller wants to query the value of such offset.
+     * After returning the value once, the redundant offset is reset.
+     * @return Optional of type Long representing the offset or empty for invalid offset values
+     */
+    public Optional<Long> lastRedundantOffset() {
```

Comment:
This method is ok, but is very customized for the usage in the current only caller. If there is another caller, it can unexpectedly break the existing caller. A better api is probably to always expose the lastRedundantOffset and let the caller handle the case when the same value is returned.